File: DP/338. Counting Bits.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> ans;
        ans.push_back(0);
        if(n==0){
            return ans;
        }
        ans.push_back(1);
        int upper=2;
        int Now=0;
        int j=0;
        for(int i=2;i<=n;i++){
            Now=ans[j]+1;
            ans.push_back(Now);
            if(Now==upper){
                upper++;
                j=0;
            }else{
                j++;
            }
        }
        return ans;
    }
};

File: DP/338. Counting Bits.java
class Solution {
    int[] arr;
    public int[] countBits(int n) {
        arr=new int[n+1];
        for(int i=n;i>0;i--){
            helper(i);
        }
        return arr;
    }
    public int helper(int n){
        if(arr[n]==0&&n!=0){
            arr[n]=helper(n>>1)+n%2;
        }
        return arr[n];
    }
}